File: src/lib/builds/assertRequiredLinks.test.ts
import { describe, expect, it } from "vitest";

import {
  assertRequiredLinksSatisfied,
  collectRequiredLinkFailures,
  equipReadyClaims,
  isRequiredLinkSatisfied,
} from "@/lib/builds/assertRequiredLinks";
import { buildInventoryPinIndex } from "@/lib/builds/equipReady";
import type { ResolvedVariantEquipment, SlotClaim } from "@/lib/builds/resolveVariant";
import type { SynergyLinkRecord, SynergyWithLinks } from "@/lib/db/repositories/synergyRepository";
import { API_ERROR_CODES, ApiError } from "@/lib/api/errors";

function link(
  partial: Partial<SynergyLinkRecord> &
    Pick<SynergyLinkRecord, "kind" | "displayName">,
): SynergyLinkRecord {
  return {
    id: partial.id ?? "L1",
    synergyId: partial.synergyId ?? "S1",
    kind: partial.kind,
    displayName: partial.displayName,
    itemHash: partial.itemHash ?? null,
    perkHash: partial.perkHash ?? null,
    parentItemHash: partial.parentItemHash ?? null,
    originTraitName: partial.originTraitName ?? null,
    originTraitHash: partial.originTraitHash ?? null,
    armorSetName: partial.armorSetName ?? null,
    bonusPieces: partial.bonusPieces ?? null,
    bonusName: partial.bonusName ?? null,
    armorSetHash: partial.armorSetHash ?? null,
    required: partial.required ?? false,
  };
}

function synergy(links: SynergyLinkRecord[]): SynergyWithLinks {
  return {
    id: "S1",
    userId: 1,
    name: "Melee: Base",
    type: "melee",
    subType: "Base",
    description: "",
    createdAt: "",
    updatedAt: "",
    links,
  };
}

function claim(
  partial: Partial<SlotClaim> & Pick<SlotClaim, "slot" | "itemHash">,
): SlotClaim {
  return {
    slot: partial.slot,
    itemHash: partial.itemHash,
    itemName: partial.itemName ?? "Item",
    source: partial.source ?? "set",
    selectedPerks: partial.selectedPerks,
    instanceId: partial.instanceId,
  };
}

function resolved(
  claims: SlotClaim[],
): ResolvedVariantEquipment {
  const equipment: ResolvedVariantEquipment["equipment"] = {};
  for (const c of claims) equipment[c.slot] = c;
  return { equipment, conflicts: [] };
}

describe("isRequiredLinkSatisfied", () => {
  it("rejects wishlist match without pin", () => {
    const l = link({
      kind: "weapon",
      displayName: "Fatebringer",
      itemHash: 10,
      required: true,
    });
    const all = [claim({ slot: "primary", itemHash: 10 })];
    const result = isRequiredLinkSatisfied(l, {
      readyClaims: [],
      allClaims: all,
    });
    expect(result).toEqual({ ok: false, reason: "wishlist_or_stale" });
  });

  it("accepts equip-ready pin match", () => {
    const l = link({
      kind: "weapon",
      displayName: "Fatebringer",
      itemHash: 10,
      required: true,
    });
    const pinned = claim({
      slot: "primary",
      itemHash: 10,
      instanceId: "inst-1",
    });
    const inv = buildInventoryPinIndex([{ instanceId: "inst-1", itemHash: 10 }]);
    const ready = equipReadyClaims(resolved([pinned]), inv);
    const result = isRequiredLinkSatisfied(l, {
      readyClaims: ready,
      allClaims: [pinned],
    });
    expect(result).toEqual({ ok: true });
  });

  it("matches artifact_perk from applied config", () => {
    const l = link({
      kind: "artifact_perk",
      displayName: "Anti-Barrier",
      perkHash: 55,
      required: true,
    });
    expect(
      isRequiredLinkSatisfied(l, {
        readyClaims: [],
        allClaims: [],
        ctx: { artifactConfig: [55] },
      }),
    ).toEqual({ ok: true });
    expect(
      isRequiredLinkSatisfied(l, {
        readyClaims: [],
        allClaims: [],
        ctx: { artifactConfig: [] },
      }),
    ).toEqual({ ok: false, reason: "unmatched" });
  });

  it("matches required aspect from applied kit without pins", () => {
    const l = link({
      kind: "aspect",
      displayName: "Roaring Flames",
      required: true,
    });
    expect(
      isRequiredLinkSatisfied(l, {
        readyClaims: [],
        allClaims: [],
        ctx: { kit: { aspects: ["Roaring Flames"] } },
      }),
    ).toEqual({ ok: true });
  });
});

describe("collectRequiredLinkFailures / assert", () => {
  it("ignores non-required links", () => {
    const failures = collectRequiredLinkFailures({
      synergies: [
        synergy([
          link({
            kind: "weapon",
            displayName: "Soft",
            itemHash: 99,
            required: false,
          }),
        ]),
      ],
      resolved: resolved([]),
      inventory: buildInventoryPinIndex([]),
    });
    expect(failures).toEqual([]);
  });

  it("throws REQUIRED_LINK_UNSATISFIED when required link missing pin", () => {
    const syn = synergy([
      link({
        kind: "weapon",
        displayName: "Required Gun",
        itemHash: 10,
        required: true,
      }),
    ]);
    try {
      assertRequiredLinksSatisfied({
        synergies: [syn],
        resolved: resolved([claim({ slot: "primary", itemHash: 10 })]),
        inventory: buildInventoryPinIndex([]),
      });
      expect.unreachable();
    } catch (e) {
      expect(e).toBeInstanceOf(ApiError);
      expect((e as ApiError).code).toBe(API_ERROR_CODES.REQUIRED_LINK_UNSATISFIED);
    }
  });

  it("passes when required link is pin-satisfied", () => {
    const syn = synergy([
      link({
        kind: "weapon",
        displayName: "Required Gun",
        itemHash: 10,
        required: true,
      }),
    ]);
    expect(() =>
      assertRequiredLinksSatisfied({
        synergies: [syn],
        resolved: resolved([
          claim({ slot: "primary", itemHash: 10, instanceId: "i1" }),
        ]),
        inventory: buildInventoryPinIndex([{ instanceId: "i1", itemHash: 10 }]),
      }),
    ).not.toThrow();
  });
});

File: flutter/packages/domain/test/assert_required_links_test.dart
import 'package:destiny2_domain/destiny2_domain.dart';
import 'package:test/test.dart';

/// Golden parity with Next `assertRequiredLinks.test.ts` (current six kinds).
void main() {
  SynergyLink link({
    required SynergyLinkKind kind,
    required String displayName,
    String id = 'L1',
    String synergyId = 'S1',
    int? itemHash,
    int? perkHash,
    bool required = false,
  }) {
    return SynergyLink(
      id: id,
      synergyId: synergyId,
      kind: kind,
      displayName: displayName,
      itemHash: itemHash,
      perkHash: perkHash,
      required: required,
    );
  }

  Synergy synergy(List<SynergyLink> links) {
    return Synergy(
      id: 'S1',
      name: 'Melee: Base',
      type: const SynergyType('melee'),
      subType: 'Base',
      links: links,
    );
  }

  SlotClaim claim({
    required EquipmentSlot slot,
    required int itemHash,
    String? instanceId,
    List<int>? selectedPerks,
  }) {
    return SlotClaim(
      slot: slot,
      itemHash: itemHash,
      itemName: 'Item',
      source: ClaimSource.set,
      instanceId: instanceId,
      selectedPerks: selectedPerks,
    );
  }

  ResolvedVariantEquipment resolved(List<SlotClaim> claims) {
    return ResolvedVariantEquipment(
      equipment: {for (final c in claims) c.slot: c},
      conflicts: const [],
    );
  }

  group('isRequiredLinkSatisfied', () {
    test('rejects wishlist match without pin', () {
      final l = link(
        kind: SynergyLinkKind.weapon,
        displayName: 'Fatebringer',
        itemHash: 10,
        required: true,
      );
      final all = [claim(slot: EquipmentSlot.primary, itemHash: 10)];
      final result = isRequiredLinkSatisfied(
        l,
        readyClaims: const [],
        allClaims: all,
      );
      expect(result.ok, isFalse);
      expect(result.reason, 'wishlist_or_stale');
    });

    test('accepts equip-ready pin match', () {
      final l = link(
        kind: SynergyLinkKind.weapon,
        displayName: 'Fatebringer',
        itemHash: 10,
        required: true,
      );
      final pinned = claim(
        slot: EquipmentSlot.primary,
        itemHash: 10,
        instanceId: 'inst-1',
      );
      final inv = buildInventoryPinIndex([
        const InventoryPinItem(instanceId: 'inst-1', itemHash: 10),
      ]);
      final ready = equipReadyClaims(resolved([pinned]), inv);
      final result = isRequiredLinkSatisfied(
        l,
        readyClaims: ready,
        allClaims: [pinned],
      );
      expect(result.ok, isTrue);
    });

    test('matches artifact_perk from applied config', () {
      final l = link(
        kind: SynergyLinkKind.artifactPerk,
        displayName: 'Anti-Barrier',
        perkHash: 55,
        required: true,
      );
      expect(
        isRequiredLinkSatisfied(
          l,
          readyClaims: const [],
          allClaims: const [],
          ctx: const MatchEvidenceContext(artifactConfig: [55]),
        ).ok,
        isTrue,
      );
      expect(
        isRequiredLinkSatisfied(
          l,
          readyClaims: const [],
          allClaims: const [],
          ctx: const MatchEvidenceContext(artifactConfig: []),
        ).ok,
        isFalse,
      );
    });
  });

  group('collectRequiredLinkFailures / assert', () {
    test('ignores non-required links', () {
      final failures = collectRequiredLinkFailures(
        synergies: [
          synergy([
            link(
              kind: SynergyLinkKind.weapon,
              displayName: 'Soft',
              itemHash: 99,
              required: false,
            ),
          ]),
        ],
        resolved: resolved(const []),
        inventory: buildInventoryPinIndex(const []),
      );
      expect(failures, isEmpty);
    });

    test('throws REQUIRED_LINK_UNSATISFIED when required link missing pin', () {
      final syn = synergy([
        link(
          kind: SynergyLinkKind.weapon,
          displayName: 'Required Gun',
          itemHash: 10,
          required: true,
        ),
      ]);
      expect(
        () => assertRequiredLinksSatisfied(
          synergies: [syn],
          resolved: resolved([
            claim(slot: EquipmentSlot.primary, itemHash: 10),
          ]),
          inventory: buildInventoryPinIndex(const []),
        ),
        throwsA(
          isA<ResolveVariantException>().having(
            (e) => e.code,
            'code',
            DomainFailureCodes.requiredLinkUnsatisfied,
          ),
        ),
      );
    });

    test('passes when required link is pin-satisfied', () {
      final syn = synergy([
        link(
          kind: SynergyLinkKind.weapon,
          displayName: 'Required Gun',
          itemHash: 10,
          required: true,
        ),
      ]);
      expect(
        () => assertRequiredLinksSatisfied(
          synergies: [syn],
          resolved: resolved([
            claim(
              slot: EquipmentSlot.primary,
              itemHash: 10,
              instanceId: 'i1',
            ),
          ]),
          inventory: buildInventoryPinIndex([
            const InventoryPinItem(instanceId: 'i1', itemHash: 10),
          ]),
        ),
        returnsNormally,
      );
    });

    test('satisfies required weapon_perk via enhanced family sibling', () {
      final family = <int, Set<int>>{
        100: {100, 101},
        101: {100, 101},
      };
      final syn = synergy([
        link(
          kind: SynergyLinkKind.weaponPerk,
          displayName: 'Zen Moment',
          perkHash: 100,
          required: true,
        ),
      ]);
      expect(
        () => assertRequiredLinksSatisfied(
          synergies: [syn],
          resolved: resolved([
            claim(
              slot: EquipmentSlot.primary,
              itemHash: 10,
              instanceId: 'i1',
              selectedPerks: const [101],
            ),
          ]),
          inventory: buildInventoryPinIndex([
            const InventoryPinItem(instanceId: 'i1', itemHash: 10),
          ]),
          ctx: MatchEvidenceContext(perkFamilyByHash: family),
        ),
        returnsNormally,
      );
    });

    test('satisfies required class-item exotic_armor via perk config', () {
      final classItems = {77};
      final syn = synergy([
        link(
          kind: SynergyLinkKind.exoticArmor,
          displayName: 'Spirit',
          itemHash: 77,
          perkHash: 900,
          required: true,
        ),
      ]);
      expect(
        () => assertRequiredLinksSatisfied(
          synergies: [syn],
          resolved: resolved([
            claim(
              slot: EquipmentSlot.classItem,
              itemHash: 77,
              instanceId: 'ci1',
              selectedPerks: const [900, 901],
            ),
          ]),
          inventory: buildInventoryPinIndex([
            const InventoryPinItem(instanceId: 'ci1', itemHash: 77),
          ]),
          ctx: MatchEvidenceContext(exoticClassItemHashes: classItems),
        ),
        returnsNormally,
      );
    });
  });
}
